### DISCUSSION TITLE
Owncast's Future and the Directory (Suggestions)

### DISCUSSION THREAD

**Discussion by andrewfer000** [owncast/owncast#1066] [Ideas]
Owncast is such an incredible piece of software. Truly Free, Open Source, and easy to use software for streaming is what I have been waiting for and now it's here and works great. The only other program that competes is OpenStreamingPlatform (OSP) but that is very complicated to setup and use. A good analogy is, Owncast is like a cupcake, the perfect size for one while OSP is like a whole cake which is good for sharing with many. Owncast Directory is like having a box of cupcakes (yum :) ) 

Owncast's main goal is to be a "Twitch in a box", as for that goal I'd say it's 80% there. Here is a small list of features I still think Owncast still needs to be "complete" on it's own. 
1. Better user security and accounts with multiple authentication methods that can be managed by admin (already in progress for 0.0.8)
2. Official documentation/support for RTMPS which already works via Nginx in Owncast 0.0.7. Support for the SRT protocol will also be a great feature to have
3. Server side stream recording and Video On Demand, which I believe is in progress for either 0.0.9 or 0.1.0 release.
4. Easy was to implement donator perks such as custom emojis and username prefixes. Subscriptions are managed through a Stripe and/or PayPal API integration.  (Maybe this is planned? I don't know) This is ESSENTIAL to be a Twitch replacement.
5. Registered users can "Follow" an instance which will then send emails to them when a stream starts. Email server is run by streamer via SMTP meaning they can run their own email server or use something like Gmail.

With those features, Owncast will pretty much be a true competitor to major platforms. But there is still one of the most important part of Owncasts growth, and that is the Directory.

The Directory provides a central hub for viewers to find instances and I think that is the most important key to getting more people on board. While one of Owncasts goals is to be decentralized. I think a Hybrid option with the directory will be prefect. Here are some things I think the directory can have to improve usability.
1. Centralized user registration and authentication through the directory can be used to authenticate on instances connected to the directory. This makes things easier for the viewer and streamer.
2. Search functionality for instances, stream names, and Video On Demand when implemented,
3. Easy subscription and follow system like what i described for the individual instances but a more centralized version that will make it easier and safer for viewers to manage their follows/subscriptions and easier for streamers to get setup and paid without having to deal with APIs or setup if they don't want to.  Owncast can also take a small cut to support the project and the directory. This is ESSENTIAL to be a Twitch replacement AND to keep the directory afloat. 
4. An Open Source "Lite version" of the directory server to allow groups to easily host events without clogging up the main directory. This lite version will focus on user authentication, following, stream/video listing, and customization.   

I'm no developer so if this all seems like too much to implement it's up to you. But the way I see it these items will truly make Owncast easy, convenient, and safe enough for both viewers and streamers to be a full replacement for Twitch. Let me know what you think.

**Comment by controlfreakstudio** [comment]


**Reply by andrewfer000** [reply]
Well this is why the directory is optional. If someone want's to use the directory they can register their instance with it and if not they can stay unlisted. This allows the people who want the convenience of centralization to have it while the people who don't want it can do what they want. Like I said, these are just simple suggestions, I do not expect them all to happen exactly like this or at all. But I feel like setting up the directory in a way like this allows the instance owner to choose to either be independent or to be put into the directory and have it's features. The admin panel can also have options to toggle these directory services on and off which I think would be the case if these features are implemented. 

Also when it comes to Subscriptions, it will be optional yet it will be set up in a way where it can be configured from the web interface even by inexperienced users. It is optional if you want it or not but it's there and ready to go. Donations are very important to full-time streamers and this is a great way to do it. Owncast already has most of the parts for it already such as integrations but it just needs a specific admin panel module and button on the web interface.

**Comment by gabek** [comment]
Thanks so much for your detailed thoughts.  I really appreciate you taking time to share your ideas.

My general mindset when it comes to Owncast continues to be "APIs over Features", meaning we could build one very specific feature or integration directly into Owncast, or we could build APIs and hooks to allow people to build anything.

> An Open Source "Lite version" of the directory server

It's something we've talked about, and it's a neat idea.  But that's a whole other project.  And admittedly the number of people actively working on Owncast are very few, so there's just no bandwidth to build another project.  If a group of people wanted to step up and focus on just the Owncast directory and build a deployable version of it I think that'd be pretty cool.  But right now the directory is a side project I  threw together as a little addition for people to discover Owncast instances, but it has never been a focus, and I don't see that changing because I can't focus on it.

> users can "Follow" an instance which will then send emails

I agree, notifications are super important for user engagement!  But in general I don't think building them into Owncast is the answer.  You mentioned email, but what if somebody else wants telegram?  What about SMS?  AOL Instant Messenger?  Instead we built a webhook that can tell any 3rd party code that a stream is live, and then you can do whatever you want with that information.  [Here's an example web app that collects phone numbers and email addresses and lets them know each time you go live](https://glitch.com/edit/#!/owncast-example-follow).  ([what it looks like to the user](https://owncast-example-follow.glitch.me/)).  I'm not saying you should use this code, but it's just an example of one way you can do what you're asking for.

> stuff about users

For each person who wants more user accounts and to need to login to things, there are just as many, or more, who want fewer accounts and don't want to deal with logging into anything.  While we're adding, very slowly, some small layers of identity, I can't imagine ever logging into the directory for anything.  I certainly don't like having more accounts for things, and I don't want others to need them either.  That's why the directory right now "just works", and there's no manual registration or extra work where you have to login and make changes.

> Subscriptions are managed through a Stripe and/or PayPal API integration.

While the "product guy" in me would love to add all kinds of native implementations for third party services, I believe this would cross the line, to build 3rd party services *into* Owncast.

If I looked at Owncast from the outside and saw it had the PayPal and Stripe SDK built in it would be a non-starter for me.  These companies monitor everything you do for "fraud detection" and to build that into Owncast directly would insult the people who love Owncast for not having things like that.  And while "easy way to make money" sounds awesome, that's giving up too much. 

So instead of building support for specific third party services I'd prefer to take a couple steps back, look wider at the problem and find ways to build hooks into Owncast for people to take advantage of.  The problem isn't "how do you use PayPal or Stripe for subscriptions in Owncast", but instead the question really is "how do we allow third party actions to take place from within an Owncast server?" and go from there, building more APIs, UI hooks, and events that can be sent and received.  And then leave it up to external people to build addons, bots, APIs and services to take advantage of these services via webhooks and inbound events. 

While I'm sure that's not what you want to hear. Having a built-in, single button way to enable monetization within Owncast would be great for some people, but that's not really what this project is all about.  We're not trying to build just twitch, or else we would have had all those things and copied twitch.  Instead we're building a platform and a piece of infrastructure to allow people to build whatever kind of streaming experience they want, and that requires more flexibility and less specificity.

I hope if people want these features they can take advantage of the APIs offered and build some cool stuff on top of them that they can offer other people.  If a group wants to build a email notification addon that people can point their owncast instance at, that would be awesome.  If somebody wants to build a notification service that alerts people via carrier pigeon, they could do that too.  (I'd really like this)

**Reply by andrewfer000** [reply]
Yes you do have a point, I believe Owncast's API and UI Hooks system is extremely innovative and very extendable. I was hoping those features  can be used in order to make ready-to-go scripts where people can just toggle them and set them up in their instance admin without having to deal with programming. I think maybe a future Owncast version should allow users to create integrations that can have a tab on the Admin panel for easy configuration. This allows people to have what they want without with the only requirement be the shadow users/user registration in 0.0.8. While I may have not mentioned it directly, the whole time I was talking about using the allready existing API and Integrations system built into Owncast but just have these specific ones here in an easy to install and setup manner.

As for the Open Source Directory thing, that can be held off for now. I think if other groups want to implement their own directory they can do it themself using the API. 

As for central logins through the directory, you got a point. No one wants more accounts if they don't want to but I thought it would just be a nice way to easily allow people who are not technically inclined to use Owncast as a primary streaming service that is user friendly to their viewers and does not take hours or days of effort to get up and running.    

By the way, what do you think of the SRT protocol and do you think it would be possible. If so it would really put Owncast in a class of it's own!
